File: 1358. number-of-substrings-containing-all-three-characters.py
class Solution:
    def numberOfSubstrings(self, s: str) -> int:
        res, p = 0, [5e4, -1, -1, -1]

        for i, ch in enumerate(s):
            p[ord(ch) & 31] = i
            res += min(p) + 1

        return res


if __name__ == "__main__":
    sol = Solution()

    print(sol.numberOfSubstrings("abcabc"))  # Expected: 10
    print(sol.numberOfSubstrings("aaacb"))   # Expected: 3
    print(sol.numberOfSubstrings("abc"))     # Expected: 1

File: 1358. number-of-substrings-containing-all-three-characters.java
class Solution {
    public int numberOfSubstrings(String s) {
        int res = 0;
        int[] p = {-1, -1, -1};

        for (int i = 0; i < s.length(); i++) {
            p[(s.charAt(i) & 31) - 1] = i;
            res += Math.min(p[0], Math.min(p[1], p[2])) + 1;
        }

        return res;
    }

    public static void main(String[] args) {
        Solution sol = new Solution();

        System.out.println(sol.numberOfSubstrings("abcabc")); // Expected: 10
        System.out.println(sol.numberOfSubstrings("aaacb"));  // Expected: 3
        System.out.println(sol.numberOfSubstrings("abc"));    // Expected: 1
    }
}